vpr/src/place/initial_placement.cpp

```diff
@@ -213,88 +231,77 @@ static void initial_placement_pl_macros(int macros_max_num_tries, std::vector<st
-        std::sort(sorted_blocks.begin(), sorted_blocks.end(), criteria);
-    }
-
-    for (auto blk_id : sorted_blocks) {
-        /* -1 is a sentinel for a non-placed block, which the code in this routine will choose a location for.
-         * If the x value is not -1, we assume something else has already placed this block and we should leave it there.
-         * For example, if the user constrained it to a certain location, the block has already been placed.
-         */
-        if (place_ctx.block_locs[blk_id].loc.x != -1) {
-            //put VTR assert
-            continue;
-        }
-
-        auto logical_block = cluster_ctx.clb_nlist.block_type(blk_id);
+            if (type == nullptr) {
+                VPR_FATAL_ERROR(VPR_ERROR_PLACE,
+                                "Initial placement failed.\n"
+                                "Could not place block %s (#%zu); no free locations of type %s (#%d).\n",
+                                cluster_ctx.clb_nlist.block_name(blk_id).c_str(), size_t(blk_id), logical_block->name, logical_block->index);
+            }
 
-        /* Randomly select a free location of the appropriate type for blk_id.
-         * We have a linearized list of all the free locations that can
-         * accommodate a block of that type in free_locations[itype].
-         * Choose one randomly and put blk_id there. Then we don't want to pick
-         * that location again, so remove it from the free_locations array.
-         */
+            int itype = type->index;
 
-        auto type = pick_placement_type(logical_block, 1, free_locations);
+            auto possible_sub_tiles = get_possible_sub_tile_indices(type, logical_block);
 
-        if (type == nullptr) {
-            VPR_FATAL_ERROR(VPR_ERROR_PLACE,
-                            "Initial placement failed.\n"
-                            "Could not place block %s (#%zu); no free locations of type %s (#%d).\n",
-                            cluster_ctx.clb_nlist.block_name(blk_id).c_str(), size_t(blk_id), logical_block->name, logical_block->index);
-        }
+            int isub_tile = get_free_sub_tile(free_locations, itype, possible_sub_tiles);
 
-        int itype = type->index;
+            t_pl_loc to;
+            int ipos = vtr::irand(free_locations[itype][isub_tile] - 1);
 
-        auto possible_sub_tiles = get_possible_sub_tile_indices(type, logical_block);
+            to = legal_pos[itype][isub_tile][ipos];
 
-        int isub_tile = get_free_sub_tile(free_locations, itype, possible_sub_tiles);
+            // Make sure that the position is EMPTY_BLOCK before placing the block down
+            VTR_ASSERT(place_ctx.grid_blocks[to.x][to.y].blocks[to.sub_tile] == EMPTY_BLOCK_ID);
 
-        t_pl_loc to;
-        int ipos = vtr::irand(free_locations[itype][isub_tile] - 1);
+            bool floorplan_good = cluster_floorplanning_check(blk_id, to);
 
-        to = legal_pos[itype][isub_tile][ipos];
+            if (floorplan_good) {
```

Comment:
found a potential location -> have a helper function that includes all the floorplan good stuff and just returns whether it was able to return or not